rfcs/20210827-benchdnn-model-support/README.md

```diff
@@ -0,0 +1,63 @@
+# Benchdnn as a new tool to measure full model performance.
+
+## Problem Statement.
+There are talks around and inside oneDNN team about the need to improve the way
+performance validation/benchmarking of full models can be done.
+
+Currently, benchdnn provides only a single way to measure performance - single
+problem is executed on same memory until certain conditions are reached. This
+allows to know a problem performance in "vacuum" environment - unrealistic
+conditions for end users. The problem of this approach that the result should be
+scaled down to be "real" but it's hard to say how much and how to estimate this
+ratio. Certain performance problems can't be found with such methodology, like
+cores trash-trafficking due to different blocking strategies in adjacent
+primitives, or data eviction from caches due to, e.g., different sets of
+weights, while a single problem fits well.
+
+The next validation step is to have oneDNN integration into a framework and run
+models they support. It may include framework overheads and contain integration
+issues which makes performance analysis harder to estimate how good the results
+are, how precise projections were and what causes the issue if targets are not
+met - oneDNN library or framework integration? Besides it also takes long time
+to integrate new features from oneDNN into framework to measure something new
+designed for performance optimizations which increases risks not to deliver the
+targeted level of performance on time.
+
+It sounds like there could be some middle state between existing modes.
+Something that could run models with oneDNN but lacking dependencies on
+framework development and overheads. It would also provide better roofline
+projections with oneDNN and performance validation on framework side with
+numbers from this new mode should be simplified. Certain issues observed on
+frameworks side only would become easier to reproduce or confirm the problem is
+in a different place. It may not solve all the problems we have now with
+performance measurements and evaluations, but it would take away some or most of
+them.
+
+## Proposal
+Let's introduce a new driver in benchdnn - "model". It would parse problems with
+their correspondent driver names, setup a problem, setup a primitive, manage
+memories, and pile them up. Once everything is parsed and stocked, run them all.
+Memories for sources would be passed from previous primitive destinations.
+
+Pros:
+* Standalone and familiar tool using oneDNN only.
+* Same syntax to trigger existing primitives, which means that most models are
+  almost ready as is.
```

Comment:
I am not sure I fully understand this statement.
- If we assume that all primitive execution are chained simply, we still need a way to delay their execution to run that chain in a loop. This would contradict the benchdnn "state-machine" model.
- if we need to support non simple graph (e.g. that contains concat or split), we would need some syntax to specify dependency between inputs/outputs to properly manage the cache residency of tensors.

Reply:
> If we assume that all primitive execution are chained simply, we still need a way to delay their execution to run that chain in a loop. This would contradict the benchdnn "state-machine" model.

We need a "state" machine only to parse the input. We parse the whole input up to the end, pile up primitives and their memories and then we execute. It won't be one-by-one execution. It also would assume/rely/validate it's a single entry for each configuration and a single descriptor passed (as in the real models). It works a bit differently compared to a single driver, but it also serves a different purpose, that's why I don't see a big problem here. If needed, may extend a mechanism with description how it is supposed to work.

> if we need to support non simple graph (e.g. that contains concat or split), we would need some syntax to specify dependency between inputs/outputs to properly manage the cache residency of tensors.

Yes, it's true. I haven't stated no new syntax will be added at all, but for linear chains it won't. It's also described in cons that branching may not make it at all. Not in the first variant for sure.

Reply:
> We need a "state" machine only to parse the input. We parse the whole input up to the end, pile up primitives and their memories and then we execute.

This option changes the execution model though, from eager to lazy. So the lazy/eagerness is part of the state. With the current proposal, this means for example that once the execution mode is switched to lazy, we cannot go back to eager.
Also, if we don't have a `stop_lazy` marker, we cannot rely on batch files to test several models in one file.

Reply:
@mgouicem Random place to provide my point of view. As kernel developer I would be very happy to measure performance of sequence of connected problems not only single problem. I think it would help, for example, in investigation and even prediction of some issues related to very different parallelization patterns in different primitives. Of course it will not fully model FW performance, especially for non-linear  models, but I think it will have much better correlation with framework performance vs just setting eviction flags for different tensors. One more thing, it might help in development and performance optimization for the future platforms as there is some latency between oneDNN and FW support enabling for the new platform, so it might help to catch some potential issues even before support in any FW

Reply:
@akharito , could you clarify how this would be helpful to find issues with different problem decomposition between primitives? Intuitively, if two consecutive primitives partition a common tensor differently, this should be shown by the amount of L2 snooping or cache line demoting in L3. So Adding an option to evict that tensor from L2 seems close enough, isn't it?

Also, let's say we limit to just chaining 2 primitives to simplify. Even in that scenario, it is hard to know where the performance drop comes from. We cannot conclude much from such a run, unless we analyze it under Vtune or some other tool to measure performance counters, and nail down the cause of the perf drop. And in that case, why not just do the same analysis on the framework run directly?

Reply:
@mgouicem unfortunately just eviction tensor from L2 can't even close reproduce one of the worst cases - when src data values for current layer required for computation on the current core are located in L2 of the other core which generates core-to-core traffic.
Yes, we can always investigate using FWK runs but this is always dependency on FWK and oneDNN support level there (which primitives, fusions, layouts propagations etc are supported there). In addition using "mimic FW" we can potentially control granularity, reducing the mumber of problems if required, while in FWK we have to run entire model... One more thing - we can potentially model performance for some cases like fusion support, usage of different primitives or layout propagation which are currently not supported in framework